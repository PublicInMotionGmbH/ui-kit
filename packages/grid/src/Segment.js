import React from 'react'
import PropTypes from 'prop-types'
import cls from 'classnames'

// As it may be used too often, class names in Grid are simplified
const name = '_'

/**
 * Component which represents Grid.
 *
 * @param {object} props
 * @param {string} [props.className]
 * @returns {React.Element}
 */
function Segment (props) {
  const { className, size, medium, large, xlarge, children, ...passedProps } = props

  const clsName = cls(name, className, {
    [`${name}-${size}`]: size,
    [`${name}-md-${medium}`]: medium,
    [`${name}-lg-${large}`]: large,
    [`${name}-xl-${xlarge}`]: xlarge
  })

  return (
    <div className={clsName} {...passedProps}>
      {children}
    </div>
  )
}

Segment.propTypes = {
  /** Additional class name */
  className: PropTypes.string,

  /** Default number of columns (from small screen) */
  size: PropTypes.number,

  /** Number of columns from medium screen */
  medium: PropTypes.number,

  /** Number of columns from large screen */
  large: PropTypes.number,

  /** Number of columns from extra-large screen */
  xlarge: PropTypes.number,

  /** Children inside a segment */
  children: PropTypes.node
}

export default Segment
