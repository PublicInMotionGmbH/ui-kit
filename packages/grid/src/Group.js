import React from 'react'
import PropTypes from 'prop-types'
import cls from 'classnames'

import Segment from './Segment'

// As it may be used too often, class names in Grid are simplified
const name = '__'

const propTypes = {
  /** Additional class name */
  className: PropTypes.string,

  /** Children inside a Group - probably some Segment (or Group/Masonry which are Segments too) */
  children: PropTypes.node,

  /** Should have spacing between */
  spaced: PropTypes.bool,

  /** Look at Segment */
  size: PropTypes.number,

  /** Look at Segment */
  medium: PropTypes.number,

  /** Look at Segment */
  large: PropTypes.number,

  /** Look at Segment */
  xlarge: PropTypes.number
}

const defaultProps = {
  spaced: false
}

/**
 * Component which represents Grid.
 *
 * @param {object} props
 * @param {string} [props.className]
 * @returns {React.Element}
 */
function Group (props) {
  const { className, spaced, children, ...passedProps } = props
  const clsName = cls(name, className, {
    [`${name}--spaced`]: !!spaced
  })

  return (
    <Segment className={clsName} {...passedProps}>
      {children}
    </Segment>
  )
}

Group.displayName = 'Group'

Group.propTypes = propTypes
Group.defaultProps = defaultProps

export default Group
