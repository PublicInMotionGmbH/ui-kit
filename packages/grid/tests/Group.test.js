import React from 'react'
import Group from '../src/Group'
import Segment from '../src/Segment'
import { shallow } from 'enzyme'

describe('<Group />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Group>
        <Segment>12</Segment>
      </Group>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('turns into Segment properly', () => {
    const wrapper = shallow(
      <Group>
        <Group size={3}>
          <Segment>3</Segment>
        </Group>
        <Group size={9}>
          <Segment>9</Segment>
        </Group>
      </Group>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('turns into Segment properly with any props', () => {
    const wrapper = shallow(
      <Group>
        <Group size={12} medium={6} large={3} xlarge={3}>
          <Segment>12/6/3</Segment>
        </Group>
        <Group size={12} medium={6} large={9} xlarge={9}>
          <Segment>12/6/9</Segment>
        </Group>
      </Group>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('handles `spaced `prop`', () => {
    const wrapper = shallow(
      <Group spaced>
        <Segment>3</Segment>
        <Segment>9</Segment>
      </Group>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('allows custom class names', () => {
    const wrapper = shallow(
      <Group className='abc'>
        <Segment>12</Segment>
      </Group>
    )

    expect(wrapper.props().className).toMatch(/(^| )abc( |$)/)
    expect(wrapper.props().className).not.toBe('abc')
  })
})
