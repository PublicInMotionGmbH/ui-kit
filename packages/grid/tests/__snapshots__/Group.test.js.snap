// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Group /> handles \`spaced \`prop\` 1`] = `
<Segment
  className="__ __--spaced"
  size={12}
>
  <Segment
    size={12}
  >
    3
  </Segment>
  <Segment
    size={12}
  >
    9
  </Segment>
</Segment>
`;

exports[`<Group /> renders correctly 1`] = `
<Segment
  className="__"
  size={12}
>
  <Segment
    size={12}
  >
    12
  </Segment>
</Segment>
`;

exports[`<Group /> turns into Segment properly 1`] = `
<Segment
  className="__"
  size={12}
>
  <Group
    size={3}
    spaced={false}
  >
    <Segment
      size={12}
    >
      3
    </Segment>
  </Group>
  <Group
    size={9}
    spaced={false}
  >
    <Segment
      size={12}
    >
      9
    </Segment>
  </Group>
</Segment>
`;

exports[`<Group /> turns into Segment properly with any props 1`] = `
<Segment
  className="__"
  size={12}
>
  <Group
    large={3}
    medium={6}
    size={12}
    spaced={false}
    xlarge={3}
  >
    <Segment
      size={12}
    >
      12/6/3
    </Segment>
  </Group>
  <Group
    large={9}
    medium={6}
    size={12}
    spaced={false}
    xlarge={9}
  >
    <Segment
      size={12}
    >
      12/6/9
    </Segment>
  </Group>
</Segment>
`;
